src/main/scala/com/databricks/labs/overwatch/MultiWorkspaceDeployment.scala

```diff
@@ -0,0 +1,369 @@
+    var configWriteLocation = configDF.head().etl_storage_prefix
+    if (!configWriteLocation.startsWith("dbfs:") && !configWriteLocation.startsWith("s3") && !configWriteLocation.startsWith("abfss")) {
+      configWriteLocation = s"""dbfs:${configWriteLocation}"""
+    }
+    configDF
+      .withColumn("snapTS", lit(pipelineSnapTime.asTSString))
+      .withColumn("timestamp", lit(pipelineSnapTime.asUnixTimeMilli))
+      .withColumn("configFile", lit(configCsvPath))
+      .write.format("delta").mode("append").option("mergeSchema", "true").save(s"""${configWriteLocation}/report/configTable""")
+  }
+
+  /**
+   * Takes a snapshot of the validation result
+   *
+   * @param validationDF
+   * @param path
+   * @param reportName
+   */
+  private def snapShotValidation(validationDF: Dataset[DeploymentValidationReport], path: String, reportName: String, deploymentId: String): Unit = {
+    var validationPath = path
+    if (!path.startsWith("dbfs:") && !path.startsWith("s3") && !path.startsWith("abfss")) {
+      validationPath = s"""dbfs:${path}"""
+    }
+    validationDF
+      .withColumn("deployment_id", lit(deploymentId))
+      .withColumn("snapTS", lit(pipelineSnapTime.asTSString))
+      .withColumn("timestamp", lit(pipelineSnapTime.asUnixTimeMilli))
+      .write.format("delta").option("mergeSchema", "true").mode("append").save(s"""${validationPath}/report/${reportName}""")
+    println("Validation report has been saved to " + s"""${validationPath}/report/${reportName}""")
+  }
+
+  /**
+   * Saves the deployment report.
+   *
+   * @param validationDF
+   * @param path
+   * @param reportName
+   */
+  private def saveDeploymentReport(validationDF: Dataset[MultiWSDeploymentReport], path: String, reportName: String): Unit = {
+    var reportPath = path
+    if (!path.startsWith("dbfs:") && !path.startsWith("s3") && !path.startsWith("abfss")) {
+      reportPath = s"""dbfs:${path}"""
+    }
+    validationDF
+      .withColumn("snapTS", lit(pipelineSnapTime.asTSString))
+      .withColumn("timestamp", lit(pipelineSnapTime.asUnixTimeMilli))
+      .write.format("delta").mode("append").save(s"""${reportPath}/report/${reportName}""")
+  }
+
+  /**
+   * Performs the multi workspace deployment
+   *
+   * @param parallelism the number of threads which will run in parallel to perform the deployment
+   * @param zones       the zone can be Bronze,Silver or Gold by default it will be Bronze
+   */
+  def deploy(parallelism: Int = 4, zones: String = "Bronze"): Unit = {
+    val processingStartTime = System.currentTimeMillis();
+    println("ParallelismLevel :" + parallelism)
+    val deploymentValidation = DeploymentValidation
+    deploymentValidation.performMandatoryValidation(configCsvPath, parallelism, deploymentId)
+    val dataFrame = deploymentValidation.makeDataFrame(configCsvPath, deploymentId)
+    snapshotConfig(dataFrame)
+    val params = dataFrame.collect().map(buildParams).filter(args => args != null).par
```

Comment:
WE can also get rid of this by already having a DS and/or a collection of Case Class derived from the DS.